prov/shm, util: various updates and fixes

aingerson opened the pull request:
General improvements to the shm provider (see specific commits for details):
- fix trimming of provider capabilities in utility code
- set shm provider tx_attr and rx_attr caps correctly
- set shm provider MR iov limit and FI_CONTEXT mode
- standardize FI_ADDR_STR usage and format in util and shm code

Signed-off-by: aingerson <alexia.ingerson@intel.com>

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -388,6 +388,8 @@ static int smr_progress_cmd_rma(struct smr_ep *ep, struct smr_cmd *cmd)
 		err = -FI_EINVAL;
 	}
+
+comp:
```
I'm not sure this is the behavior we want.  IIUC, this is the target of an RMA write which failed it's memory region check.  The receiving side should discard the operation, with only the transmit side (possibly) seeing a failure.

aingerson replied:
Got it. Then I misunderstood. I noticed this when a test was hanging instead of reporting a MR verification failure. I will throw it out then if that's not the expected behavior.

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -1162,8 +1173,7 @@ static uint64_t ofi_get_info_caps(const struct fi_info *prov_info,
 
-trim_caps:
-	return caps & ~(FI_REMOTE_WRITE | FI_REMOTE_READ);
+	return caps;
```
Unless I'm reading this function wrong, we call ofi_get_caps() and then return that value in all cases.  So, we don't really need this function at all anymore.  @a-ilango - why were the caps being trimmed, and is it still needed given the other proposed changes?

a-ilango replied:
We trim the caps here to prevent providers that require RMA target MR modes from advertising support for RMA target caps when the user doesn't support the required modes. It is still needed. The provider should set correct mr_modes for this function to work correctly.

Does shm provider require any of OFI_MR_MODE_RMA_TARGET modes? If not this function would return the caps without trimming.

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -32,8 +32,13 @@
+#define SMR_TX_CAPS (FI_MSG | FI_SEND | FI_RMA | FI_READ | 		\
+		     FI_WRITE | FI_ATOMICS)
+#define SMR_RX_CAPS (FI_MSG | FI_RECV | FI_RMA | FI_REMOTE_READ |	\
+		     FI_REMOTE_WRITE | FI_ATOMICS | FI_SOURCE)
```
I'll wager that most providers are setting the Tx/Rx caps incorrectly when it comes to FI_READ, FI_WRITE, FI_REMOTE_READ, FI_REMOTE_WRITE.  I wonder if the above two defines shouldn't be moved into common code. Maybe something like OFI_BASE_TX_CAPS?  Or maybe break it out more, e.g. OFI_TX_MSG_CAPS, OFI_TX_RMA_CAPS, etc.

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -80,8 +85,7 @@ struct fi_fabric_attr smr_fabric_attr = {
 struct fi_info smr_info = {
-	.caps = FI_MSG | FI_SEND | FI_RECV | FI_SOURCE | FI_TAGGED | FI_RMA |
-		FI_READ | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE | FI_ATOMICS,
+	.caps = SMR_TX_CAPS | SMR_RX_CAPS | FI_TAGGED,
```
FI_TAGGED should move up next with FI_MSG in the *_CAPS definitions.

shefty commented on prov/shm/src/smr.h:
```diff
@@ -161,6 +161,16 @@ struct smr_domain {
 };
 
+#define SMR_PREFIX	"fi_shm://"
+#define SMR_PREFIX_NS	"fi_ns://"
```
The commit message for this change is the most beautiful commit message I've ever read.  It nearly brought a tear to my eye.  Please insert the message as a section in the shm man page on addressing, or into some other documentation that will allow a user to understand how shm addressing works.  Then this will not be buried in the git history abyss.

aingerson replied:
I thought you would enjoy that! I haven't updated the man page since the original commit. Now that it's pretty stable and functional, I'm working on updating the man pages with a more accurate description of the provider. I'll make sure to include this commit message in that update.

shefty commented on prov/shm/src/smr.h:
```diff
@@ -161,6 +161,16 @@ struct smr_domain {
+#define SMR_PREFIX	"fi_shm://"
+#define SMR_PREFIX_NS	"fi_ns://"
+
+static inline char *smr_no_prefix(char *addr)
```
I think the function and parameter should both be const.

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -382,18 +382,18 @@ static struct fi_ops smr_ep_fi_ops = {
 {
+	char *start;
 	memset(name, 0, SMR_NAME_SIZE);
-	if (addr) {
```
The change loses this check against a null addr.  Do all callers provide a non-null address?

aingerson replied:
The addr comes from info->src_addr which is guaranteed to be non-null after getinfo with the correct address. Unless the user frees src_addr, it should be ok to drop the check, but it might be safer to keep it after all.

aingerson commented:
Updated.
- Got rid of RX completion commit. 
- Added setting FI_CONTEXT mode to mr_iov_limit commit
- Removed the removal of the extra FI_REMOTE_READ and FI_REMOTE_WRITE trim and added a fix to an incorrect MR mode check where it wasn't returning when it should have

a-ilango commented on prov/util/src/util_attr.c:
```diff
@@ -1158,8 +1169,7 @@ static uint64_t ofi_get_info_caps(const struct fi_info *prov_info,
 	    (user_mode == FI_MR_BASIC) ||
-	    ((user_mode & OFI_MR_MODE_RMA_TARGET) ==
-	     (prov_mode & OFI_MR_MODE_RMA_TARGET)))
+	    ((user_mode & prov_mode) & OFI_MR_MODE_RMA_TARGET))
```
Shouldn't this be `(user_mode & prov_mode & OFI_MR_MODE_RMA_TARGET) == (prov_mode & OFI_MR_MODE_RMA_TARGET)`?

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -1041,6 +1044,14 @@ static uint64_t ofi_get_caps(uint64_t info_caps, uint64_t hint_caps,
 		       (attr_caps & FI_SECONDARY_CAPS);
 	}
+
+	if (caps & FI_MSG && !(caps & OFI_MSG_CAPS))
```
I think this needs to be:
```
if ((caps & (FI_MSG | FI_TAGGED)) && !(caps & OFI_MSG_CAPS))
```

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -1041,6 +1044,14 @@ static uint64_t ofi_get_caps(uint64_t info_caps, uint64_t hint_caps,
+		caps |= OFI_MSG_CAPS;
+	if (caps & FI_RMA && !(caps & OFI_RMA_CAPS))
+		caps |= OFI_RMA_CAPS;
+	if (caps & FI_ATOMICS && !(caps & OFI_RMA_CAPS))
```
This can be combined with the previous check
```
if ((caps & (FI_RMA | FI_ATOMIC)) && !(caps & OFI_RMA_CAPS))
```

shefty commented on prov/util/src/util_attr.c:
```diff
@@ -1158,7 +1169,7 @@ static uint64_t ofi_get_info_caps(const struct fi_info *prov_info,
 	    (user_mode == FI_MR_UNSPEC)) ||
 	    (user_mode == FI_MR_BASIC) ||
-	    ((user_mode & OFI_MR_MODE_RMA_TARGET) ==
+	    ((user_mode & prov_mode & OFI_MR_MODE_RMA_TARGET) &&
```
Did you mean '==' here instead of '&&'?

aingerson replied:
Yes, I did

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -85,6 +86,7 @@ struct fi_fabric_attr smr_fabric_attr = {
 
 struct fi_info smr_info = {
 	.caps = SMR_TX_CAPS | SMR_RX_CAPS,
+	.mode = FI_CONTEXT,
```
Does shm require FI_CONTEXT? I didn't think that it did.

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -80,8 +85,8 @@ struct fi_fabric_attr smr_fabric_attr = {
-	.caps = FI_MSG | FI_SEND | FI_RECV | FI_SOURCE | FI_TAGGED | FI_RMA |
-		FI_READ | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE | FI_ATOMICS,
+	.caps = SMR_TX_CAPS | SMR_RX_CAPS,
+	.mode = FI_CONTEXT,
```
I don't see where this mode bit is needed.  Setting this requires that the app provide struct fi_context as its operation context parameter.  From what I can tell, the shm provider doesn't care what the user passes in as its context.   Nothing in this series needs it.

aingerson replied:
The provider needs a valid context in the case of larger messages to track the pending command and report errors after finding a completed response. It also needs a context for receives that might be canceled (which I'm working on right now).

shefty replied:
I would drop this until the patch that needs it is added.  Adding mode bits reduces the number of apps that can use the provider.  In the worst case, we allocate a context internally if needed.

aingerson commented:
Merging as is, with the FI_CONTEXT change. I'll take it out in a follow up commit when it is no longer officially needed.
